src/memory_limit.rs

```diff
@@ -0,0 +1,182 @@
+// Copyright 2017, 2018 Parity Technologies
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{Decode, Error, Input};
+
+/// The error message returned when the memory limit is reached.
+const DECODE_OOM_ERROR: &str = "Out of memory when decoding";
+
+/// Extension trait to [`Decode`] for decoding with a maximum memory consumption.
+///
+/// Should be used as a marker for types that do the memory tracking in their `decode` implementation with [`Input::try_alloc`].
+pub trait DecodeMemLimit: Decode + Sized {
+	fn decode_with_mem_limit<I: Input>(limit: MemLimit, input: &mut I) -> Result<Self, Error> {
```

Comment:
This approach seems a bit limiting since we can only do `decode_with_mem_limit` or `decode_with_depth_limit`. Looks like we can't combine them. It would be nice to be able to do something more generic. For example something like:

`decode_with_check(input: &mut I, check: FnMut(DecodeContext) -> Result<(), Error>)` 

where `DecodeContext` would be something like

```
DecodeContext {
    depth: u32,
    used_memory: usize,
    ...
}
```

This way we could check both for depth and memory limit at the same time. Maybe for other parameters as well. Also would be nice to be able to keep track of the number of objects of a certain type that we encounter recursively while decoding since this is also something that we needed in Polkadot.

Reply:
> Looks like we can't combine them. 

We can. Which is why i chose this approach. These functions are just wrappers, but there is nobody stopping you from doing:

```rust
let mut input = MemTrackingInput { inner: DepthTrackingInput { inner: .. }}
```

>This way we could check both for depth and memory limit at the same time. Maybe for other parameters as well. Also would be nice to be able to keep track of the number of objects of a certain type that we encounter recursively while decoding since this is also something that we needed in Polkadot.

Yea, i had some `decode_with_context<C>(..)` implemented before, but it requires adapting all implementations of the type and forced them to use a specific `Context` type, unless we make the context a trait, but that clusterfucks everything.